lib/regexp/regexp.go

```diff
@@ -0,0 +1,290 @@
+	)
+
+	if err := starlark.UnpackPositionalArgs(b.Name(), args, kwargs, 1, &src, &max); err != nil {
+		return nil, err
+	}
+
+	re := b.Receiver().(*Regexp).re
+	return toList(re.FindAllString(src, max)), nil
+}
+
+func findSubmatches(_ *starlark.Thread, b *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	var src string
+
+	if err := starlark.UnpackPositionalArgs(b.Name(), args, kwargs, 1, &src); err != nil {
+		return nil, err
+	}
+
+	re := b.Receiver().(*Regexp).re
+	return toList(re.FindStringSubmatch(src)), nil
+}
+
+func replaceAll(thread *starlark.Thread, b *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	var (
+		src  string
+		repl starlark.Value
+	)
+
+	if err := starlark.UnpackPositionalArgs(b.Name(), args, kwargs, 2, &src, &repl); err != nil {
+		return nil, err
+	}
+
+	re := b.Receiver().(*Regexp).re
+	switch x := repl.(type) {
+	case starlark.Callable:
+		var fnErr error
+		fn := func(matched string) string {
+			res, err := starlark.Call(thread, repl, starlark.Tuple{starlark.String(matched)}, nil)
+			if err != nil {
+				// Save the error to be able to return it to the caller
+				fnErr = fmt.Errorf("%s: An error occured while executing the callable: %s", x.Name(), err.Error())
+				return ""
+			}
+			resp, ok := res.(starlark.String)
+			if !ok {
+				// Save the error to be able to return it to the caller
+				fnErr = fmt.Errorf("%s: A string is expected as return type of the callable but was %s", x.Name(), res.Type())
+				return ""
+			}
+			return string(resp)
+		}
+		result := re.ReplaceAllStringFunc(src, fn)
+		if fnErr != nil {
+			return nil, fnErr
+		}
+		return starlark.String(result), nil
+	case starlark.String:
+		return starlark.String(re.ReplaceAllString(src, convertReplacementPattern(string(x)))), nil
+	}
+	return nil, fmt.Errorf("got %s, want a string or callable", repl.Type())
+}
+
+func split(_ *starlark.Thread, b *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	var (
+		src string
+		max int = -1
+	)
+
+	if err := starlark.UnpackPositionalArgs(b.Name(), args, kwargs, 1, &src, &max); err != nil {
+		return nil, err
+	}
+
+	re := b.Receiver().(*Regexp).re
+	return toList(re.Split(src, max)), nil
+}
+
+// Converts the replacement pattern to make it compatible with what the regexp package expects.
+func convertReplacementPattern(repl string) string {
+	// Escapes the specifal dollar characters if any
+	repl = strings.ReplaceAll(repl, "$", "$$")
+	startIdx := 0
+	var sb strings.Builder
+	// Replaces the backreferences represented by backslash-escaped digits (\1 to \9) with their
+	// counter parts (${1} to ${9}).
+	for _, match := range backreferenceRe.FindAllStringSubmatchIndex(repl, -1) {
+		startIdxMatch := match[0]
+		endIdxMatch := match[1]
+		if startIdxMatch > 0 && repl[startIdxMatch-1] == '\\' {
+			// The preceding character is a slash so we don't have a real match let's skip it
```

Comment:
This too is not sound: the digit may be preceded by two backslashes. You can't do these kinds of lexical analysis correctly unless you make a complete pass from the beginning.

Reply:
What you describe: digit preceded by an even amount of backslashes is typically the reason why I added this test. If you check well, your usecase is already covered in the test cases https://github.com/google/starlark-go/pull/369/files#diff-bebff646c9e370f4768b2380210912756b9fb8ff362d777fbfcab94694ada0ceR75-R77